### PR TITLE
Update MaxPool2D to Support Width and Block Sharding

### PR DESCRIPTION
**Pull request by wransom-TT**
### Ticket
#13077 
#13078 

### Problem description
This PR updates MaxPool2D to support both width and block sharding.

### What's changed
A default parameter `applied_shard_scheme` has been added to `ttnn.max_pool2d`.  When not specified, MaxPool2D will use height sharding.  But, if either width or block sharding is specified, MaxPool2D will use that shard scheme instead.

This was primarily achieved by:
- scaling the references to the channel dimensions of the activation matrix in the program factory by the number of shards in the channel dimension
- adjusting the sliding window config calculation to pad the top left index shards to ensure all cores' shards have uniform length
- fixing a bug in `determine_parallel_conig` to compute the number of NHW cores variably depending on that shard orientation

### Checklist
- [x] Post commit CI passes
- [N/A] Blackhole Post commit (if applicable)
- [N/A] Model regression CI testing passes (if applicable)
- [N/A] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_device_op.cpp [hunk 1]
```diff
@@ -26,10 +26,16 @@ void validate_maxpool(const Tensor& input, const sliding_window::SlidingWindowCo
     TT_FATAL(is_pow2, "Row size (nchannels * bytes = {}) should be power of 2 ({}).", in_nbytes_c, is_pow2);
 
     TT_FATAL(input.memory_config().is_sharded(), "Input needs to be sharded");
-    TT_FATAL(input.memory_config().memory_layout == TensorMemoryLayout::HEIGHT_SHARDED, "Only height sharded tensors are supported.");
+    TT_FATAL((input.memory_config().memory_layout == TensorMemoryLayout::HEIGHT_SHARDED) ||
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_device_op.cpp:29]: this is now redundant with the above check is_sharded

Comment (pavlejosipovic) [reply ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_device_op.cpp:29]: same case for test below

File: ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_multi_core_program_factory.cpp [hunk 2]
```diff
@@ -49,20 +51,27 @@ MaxPool2D::MultiCore::cached_program_t max_pool_2d_multi_core_sharded_with_halo_
...
-        (out_nbytes_c & (out_nbytes_c - 1)) == 0, "out_nbytes_c should be power of 2");  // out_nbytes_c is power of 2
+    TensorMemoryLayout in_memory_layout = input.memory_config().memory_layout;
+
+    TT_FATAL((in_memory_layout == TensorMemoryLayout::HEIGHT_SHARDED) ||
+             (input_shape[3] % num_shards_c == 0), "For width sharding, input channels should be divisible by num_shards");
+    TT_FATAL((in_memory_layout == TensorMemoryLayout::HEIGHT_SHARDED) ||
+             (output_shape[3] % num_shards_c == 0), "For width sharding, output channels should be divisible by num_shards");
+    uint32_t in_nbytes_c = input_shape[3] / num_shards_c * in_nbytes;                                      // row of input (channels)
+    uint32_t out_nbytes_c = output_shape[3] / num_shards_c * out_nbytes;                                // row of output (channels)
+    //TT_ASSERT((in_nbytes_c & (in_nbytes_c - 1)) == 0, "in_nbytes_c should be power of 2");  // in_nbytes_c is power of 2
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_multi_core_program_factory.cpp:62]: why is this code under comment?

File: ttnn/cpp/ttnn/operations/pool/maxpool/max_pool2d.cpp [hunk 3]
```diff
@@ -35,8 +44,14 @@ Tensor MaxPool2DOp::invoke(uint8_t queue_id, const Tensor& input_tensor, uint32_
 
     if (!out_memory_config.shard_spec.has_value()) {
         // Input is not sharded. Perform sharding.
+        applied_shard_scheme = applied_shard_scheme == TensorMemoryLayout::INTERLEAVED ? TensorMemoryLayout::HEIGHT_SHARDED : applied_shard_scheme; // default to height sharding
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/pool/maxpool/max_pool2d.cpp:47]: I think we will want a heurisic to pick best sharding based on tesor dims here instead of just picking HS.
In our ttnn torch integration, compiler passes in just interleaved tensors and we need to make sure that we pick the best variant
to avoid out of memory issues

File: ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_multi_core_program_factory.cpp [hunk 4]
```diff
@@ -49,20 +51,27 @@ MaxPool2D::MultiCore::cached_program_t max_pool_2d_multi_core_sharded_with_halo_
...
     uint32_t in_nbytes = datum_size(in_df);
     uint32_t out_nbytes = datum_size(out_df);
-    uint32_t in_nbytes_c = input_shape[3] * in_nbytes;                                      // row of input (channels)
-    uint32_t out_nbytes_c = output_shape[3] * out_nbytes;                                   // row of output (channels)
-    TT_ASSERT((in_nbytes_c & (in_nbytes_c - 1)) == 0, "in_nbytes_c should be power of 2");  // in_nbytes_c is power of 2
-    TT_ASSERT(
-        (out_nbytes_c & (out_nbytes_c - 1)) == 0, "out_nbytes_c should be power of 2");  // out_nbytes_c is power of 2
+    TensorMemoryLayout in_memory_layout = input.memory_config().memory_layout;
+
+    TT_FATAL((in_memory_layout == TensorMemoryLayout::HEIGHT_SHARDED) ||
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_multi_core_program_factory.cpp:56]: should check like this be in validate function?

File: tests/ttnn/unit_tests/operations/test_maxpool2d.py [hunk 5]
```diff
@@ -153,6 +153,205 @@ def run_max_pool(
...
+    isclose = torch.all(torch.isclose(output_pytorch, golden_pytorch, atol=atol))
+    isequal = torch.equal(output_pytorch, golden_pytorch)
+
+    assert allclose
+    assert isclose
+    if dtype == ttnn.bfloat16:
+        assert isequal
+
+
+def run_max_pool_block_shard(
```
Comment (pavlejosipovic) [inline tests/ttnn/unit_tests/operations/test_maxpool2d.py:252]: run_max_pool_block_shard and run_max_pool_width_shard seems very similar can we merge them?

**Review by pavlejosipovic — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/test_maxpool2d.py [hunk 6]
```diff
@@ -538,3 +350,507 @@ def test_pool_core_nondivis(
...
+#         atol = 0.35
+
+#     allclose = torch.allclose(output_pytorch, golden_pytorch, atol=atol)
+#     isclose = torch.all(torch.isclose(output_pytorch, golden_pytorch, atol=atol))
+#     isequal = torch.equal(output_pytorch, golden_pytorch)
+
+#     assert allclose
+#     assert isclose
+#     if dtype == ttnn.bfloat16:
+#         assert isequal
```
Comment (mywoodstock) [inline tests/ttnn/unit_tests/operations/test_maxpool2d.py:856]: remove all commented out code?

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp [hunk 7]
```diff
@@ -95,6 +95,7 @@ ParallelConfig determine_parallel_config(
         conv_out_2d_matrix_width_ntiles = output_channels;
     }
     auto compute_with_storage_grid_size = device->compute_with_storage_grid_size();
+    printf("grid dims: %ld, %ld\n", compute_with_storage_grid_size.x, compute_with_storage_grid_size.y);
```
Comment (mywoodstock) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:98]: ensure to remove the print statements, or change them to `log_debug`.

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp [hunk 8]
```diff
@@ -136,6 +137,10 @@ ParallelConfig determine_parallel_config(
                 block_shard_orientation == ShardOrientation::COL_MAJOR ? num_cores_nhw : num_cores_channels;
             uint32_t cores_y =
                 block_shard_orientation == ShardOrientation::COL_MAJOR ? num_cores_channels : num_cores_nhw;
+                printf("total_cores_for_channels: %d\n", total_cores_for_channels);
+                printf("num_cores_channels: %d\n", num_cores_channels);
+                printf("cores_x: %d\n", cores_x);
+                printf("cores_y: %d\n", cores_y);
```
Comment (mywoodstock) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:143]: remove prints

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp [hunk 9]
```diff
@@ -162,9 +167,12 @@ uint32_t get_num_cores_nhw_from_parallel_config(const ParallelConfig& pconfig) {
     auto grid_size = pconfig.grid.bounding_box().grid_size();
     uint32_t num_cores = pconfig.grid.num_cores();
     uint32_t num_cores_nhw = 0;
-    if(pconfig.shard_scheme == TensorMemoryLayout::WIDTH_SHARDED) {
+    if (pconfig.shard_scheme == TensorMemoryLayout::WIDTH_SHARDED) {
         return 1;
     }
+    printf("grid_size.x: %ld\n", grid_size.x);
+    printf("grid_size.y: %ld\n", grid_size.y);
+    printf("num_cores: %d\n", num_cores);
```
Comment (mywoodstock) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:175]: remove prints

File: ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_multi_core_program_factory.cpp (hunk 4 above)
Comment (mywoodstock) [reply ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_multi_core_program_factory.cpp:56]: yeah all input tensor validation should go to validate function.

File: ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_multi_core_program_factory.cpp (hunk 2 above)
Comment (mywoodstock) [reply ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_multi_core_program_factory.cpp:62]: @wransom-TT Actually keep these for now -- i'll remove them in a PR im working on.

File: ttnn/cpp/ttnn/operations/pool/maxpool/max_pool2d.cpp (hunk 3 above)
Comment (mywoodstock) [reply ttnn/cpp/ttnn/operations/pool/maxpool/max_pool2d.cpp:47]: @pavlejosipovic more work for your heuristics :-D Lets do that in a subsequent PR.

File: ttnn/cpp/ttnn/operations/pool/maxpool/max_pool2d.cpp [hunk 10]
```diff
@@ -47,23 +62,38 @@ Tensor MaxPool2DOp::invoke(uint8_t queue_id, const Tensor& input_tensor, uint32_
                                             false);
         num_cores_nhw = conv::conv2d::get_num_cores_nhw_from_parallel_config(parallel_config);
         auto sharded_mem_config = conv::conv2d::create_sharded_memory_config_from_parallel_config(input_tensor_sharded.shape(), parallel_config, is_in_tiled ? tt::constants::TILE_HEIGHT : 1);
-        input_tensor_sharded = ttnn::to_memory_config(input_tensor_sharded, sharded_mem_config, std::nullopt);
+        printf("shard spec shape: %d, %d\n", sharded_mem_config.shard_spec.value().shape[0], sharded_mem_config.shard_spec.value().shape[1]);
+        input_tensor_sharded = ttnn::to_memory_config(input_tensor_sharded, sharded_mem_config, std::nullopt); // this converts interleaved to sharded
+        printf("WE DON'T GET TO HERE\n");
```
Comment (mywoodstock) [inline ttnn/cpp/ttnn/operations/pool/maxpool/max_pool2d.cpp:67]: remove prints

File: ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp [hunk 11]
```diff
@@ -371,18 +371,17 @@ std::tuple<std::vector<std::vector<uint16_t>>, std::vector<std::vector<uint16_t>
...
-std::vector<std::vector<uint16_t>> generate_sliding_window_op_config(const std::vector<uint32_t>& op_trace_metadata, const std::vector<std::pair<uint32_pair_t, uint32_pair_t>>& shard_boundaries, bool pad_tile, bool pad_last_core) {
+std::vector<std::vector<uint16_t>> generate_sliding_window_op_config(const std::vector<uint32_t>& op_trace_metadata, const std::vector<std::pair<uint32_pair_t, uint32_pair_t>>& shard_boundaries, bool pad_tile, bool pad_cores) {
     std::vector<std::vector<uint16_t>> sharded_input_top_left_indices;
     for(const auto& item : shard_boundaries) {
         const auto& [output_shard_start, output_shard_end] = item.first;
         const auto& [input_shard_start, input_shard_end] = item.second;
+        std::vector<uint16_t> local_top_left_indices;
         // sanity check
         if (output_shard_start >= op_trace_metadata.size()) {
-            // this core has no output
```
Comment (mywoodstock) [inline ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp:381]: keep the comment?

**Review by mywoodstock — CHANGES_REQUESTED** [review]

File: ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_multi_core_program_factory.cpp [hunk 12]
```diff
@@ -394,13 +403,14 @@ MaxPool2D::MultiCore::cached_program_t MaxPool2D::MultiCore::create(const operat
...
 
+    bool is_width_sharded = input.memory_config().memory_layout == TensorMemoryLayout::WIDTH_SHARDED;
     bool is_block_sharded = input.memory_config().memory_layout == TensorMemoryLayout::BLOCK_SHARDED;
 
     auto pad_metadata = sliding_window::generate_pad_metadata(sliding_window_config);
     auto op_trace_metadata = sliding_window::generate_op_trace_metadata(sliding_window_config);
     auto shard_boundaries = sliding_window::generate_shard_boundaries(sliding_window_config, op_trace_metadata);
     auto top_left_indices =
-        sliding_window::generate_sliding_window_op_config(op_trace_metadata, shard_boundaries, false, false);
+        sliding_window::generate_sliding_window_op_config(op_trace_metadata, shard_boundaries, false, is_block_sharded); // should we always pad cores or only for block sharding?
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_multi_core_program_factory.cpp:413]: Please either figure out the answer and write the code appropriately or make the comment in the form of a TODO for easier searchability.

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp [hunk 13]
```diff
@@ -106,7 +107,7 @@ ParallelConfig determine_parallel_config(
         if(shard_layout == TensorMemoryLayout::HEIGHT_SHARDED) {
             return find_closest_largest_divisor(conv_out_2d_matrix_height_ntiles, max_num_cores);
         } else if(shard_layout == TensorMemoryLayout::BLOCK_SHARDED) {
-            return find_closest_largest_divisor_with_num_padding(conv_out_2d_matrix_height_ntiles, device_grid_size[0]);
+            return find_closest_largest_divisor_with_num_padding(conv_out_2d_matrix_height_ntiles, device_grid_size[1]);
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:110]: Why the change?

**Review by bbradelTT — COMMENTED** [review]
Now that the PR is no longer a draft, please update the description of the PR to describe the changes.

**Comment by bbradelTT** [comment]
> Now that the PR is no longer a draft, please update the description of the PR to describe the changes.

Never mind. I saw that this is done already.

File: tests/ttnn/unit_tests/operations/test_maxpool2d.py [hunk 14]
```diff
@@ -153,6 +153,205 @@ def run_max_pool(
...
 
+def run_max_pool_width_shard(
+    act_shape,
+    kernel_size,
+    padding,
+    stride,
+    dilation,
+    device,
+    dtype,
+):
```
[the remark this answers is not in the document]
Comment (wransom-TT) [reply tests/ttnn/unit_tests/operations/test_maxpool2d.py:164]: Oh I see, I misunderstood when you said to convert these to separate functions, I will adjust

File: tests/ttnn/unit_tests/operations/test_maxpool2d.py (hunk 5 above)
Comment (wransom-TT) [reply tests/ttnn/unit_tests/operations/test_maxpool2d.py:252]: will do

File: tests/ttnn/unit_tests/operations/test_maxpool2d.py [hunk 15]
```diff
@@ -249,6 +448,123 @@ def test_run_max_pool_mem_config(
     run_max_pool(act_shape, (3, 3), (1, 1), (2, 2), (1, 1), device, ttnn.bfloat16, memory_config=memory_config)
 
 
+@pytest.mark.parametrize("device_params", [{"l1_small_size": 24576}], indirect=True)
+@pytest.mark.parametrize(
+    "act_shape",  ## NCHW
+    (
+        (  ## resnet shapes
```
[the remark this answers is not in the document]
Comment (wransom-TT) [reply tests/ttnn/unit_tests/operations/test_maxpool2d.py:455]: will remove!

File: ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_device_op.cpp (hunk 1 above)
Comment (wransom-TT) [reply ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_device_op.cpp:29]: got it will remove!

File: ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_multi_core_program_factory.cpp (hunk 4 above)
Comment (wransom-TT) [reply ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_multi_core_program_factory.cpp:56]: sounds good will do

File: ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp (hunk 11 above)
Comment (wransom-TT) [reply ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp:381]: oh sorry I had added functionality here which I then removed, I will add the comment back

File: ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_multi_core_program_factory.cpp (hunk 12 above)
Comment (wransom-TT) [reply ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_multi_core_program_factory.cpp:413]: The answer is we should only do this for block sharding, comment has been updated!

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp (hunk 13 above)
Comment (wransom-TT) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp]: This is outdated, but a similar change has been made in the updated version.  This was a bug in `determine_parallel_config`.  The divisor needs to be either index 1 or 0 depending on whether the shard orientation is column or row major.

File: ttnn/cpp/ttnn/operations/pool/maxpool/max_pool2d_pybind.cpp [hunk 16]
```diff
@@ -20,23 +20,24 @@ void bind_max_pool2d_operation(py::module& module) {
...
+        | kernel_h             | kernel window height          | uint32_t           |             | Yes      |
+        | kernel_w             | kernel window width           | uint32_t           |             | Yes      |
+        | stride_h             | stride in height dim          | uint32_t           |             | No       |
+        | stride_w             | stride in width dim           | uint32_t           |             | No       |
+        | pad_h                | padding in height dim         | uint32_t           |             | No       |
+        | pad_w                | padding in width dim          | uint32_t           |             | No       |
+        | dilation_h           | kernel dilation in height dim | uint32_t           |             | No       |
+        | dilation_w           | kernel dilation in width dim  | uint32_t           |             | No       |
+        | memory_config        | Output memory config          | MemoryConfig       |             | No       |
+        | applied_shard_scheme | Shard scheme                  | TensorMemoryLayout |             | No       |
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/pool/maxpool/max_pool2d_pybind.cpp:39]: It would be good to add a description (following proper doc layout) of what max pool2d does and which combinations of parameters are valid. E.g. what does applied shard scheme do, when should applied shard scheme be specified.

Also, why is interleaved chosen as the default instead of using std::optional?

**Review by bbradelTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/pool/maxpool/max_pool2d_pybind.cpp (hunk 16 above)
Comment (wransom-TT) [reply ttnn/cpp/ttnn/operations/pool/maxpool/max_pool2d_pybind.cpp:39]: Borys, I have updated MaxPool2D to use std::optional instead of the default parameter and I think this results in much clearer logic!

I was wondering if you could point me to either the pybind doc layout SOP or a good example I might be able to reference and I can add a description here.

Comment (bbradelTT) [reply ttnn/cpp/ttnn/operations/pool/maxpool/max_pool2d_pybind.cpp:39]: You can use ttnn/cpp/ttnn/operations/matmul/matmul_pybind.cpp

**Review by mywoodstock — APPROVED** [review]
nice work!

**Review by TT-BrianLiu — APPROVED** [review]
Minor comments

File: ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_multi_core_program_factory.cpp [hunk 17]
```diff
@@ -394,13 +397,14 @@ MaxPool2D::MultiCore::cached_program_t MaxPool2D::MultiCore::create(const operat
     uint32_t out_h = output_shape[1];
     uint32_t out_w = output_shape[2];
 
+    bool is_width_sharded = input.memory_config().memory_layout == TensorMemoryLayout::WIDTH_SHARDED;
```
[the remark this answers is not in the document]
Comment (wransom-TT) [reply ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_multi_core_program_factory.cpp:400]: Brian, I will address these in a following PR based off this branch.

File: ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_device_op.cpp [hunk 18]
```diff
@@ -26,10 +26,15 @@ void validate_maxpool(const Tensor& input, const sliding_window::SlidingWindowCo
...
-
     TT_FATAL(out_mem_config.is_sharded(), "Output memory config needs to be sharded");
-    TT_FATAL(out_mem_config.memory_layout == TensorMemoryLayout::HEIGHT_SHARDED, "Only height sharded tensors are supported.");
+
+    // check that C dimnenion is a multiple of num_shards_c for all but height sharding
+    TensorMemoryLayout in_memory_layout = input.memory_config().memory_layout;
+    if (in_memory_layout != TensorMemoryLayout::HEIGHT_SHARDED) {
+        uint32_t num_shards_c = sliding_window_config.num_cores_c;
+        const tt::tt_metal::LegacyShape input_shape = input.get_legacy_shape();
+        TT_FATAL(input_shape[3] % num_shards_c == 0, "For width and block sharding, input channels should be divisible by num_shards");
```
[the remark this answers is not in the document]
Comment (wransom-TT) [reply ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_device_op.cpp:36]: Brian, I will address these in a following PR based off this branch.

[the remark this answers is not in the document]
Comment (TT-BrianLiu) [reply ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_device_op.cpp:36]: Why not just fix them in this PR? This is trivial

File: ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_multi_core_program_factory.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (TT-BrianLiu) [reply ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_multi_core_program_factory.cpp:400]: Same comment. Please fix in same PR next time
